numpy/core/tests/test_numeric.py

```diff
@@ -1254,6 +1254,37 @@ def __bool__(self):
         a = np.array([[False], [TrueThenFalse()]])
         assert_raises(RuntimeError, np.nonzero, a)
 
+    def test_nonzero_exception_safe(self):
+        # gh-13930
+
+        class ThrowsAfter:
+            def __init__(self, iters):
+                ThrowsAfter.iters_left = iters
+
+            def __bool__(self):
+                if ThrowsAfter.iters_left == 0:
+                    raise ValueError("called `iters` times")
+                ThrowsAfter.iters_left -= 1
+                return True
+
+        """
+        Test that a ValueError is raised instead of a SystemError
+
+        If the __bool__ function is called after the error state is set,
+        Python (cpython) will raise a SystemError.
+        """
+
+        # assert that an exception in first pass is handled correctly
+        a = np.array([ThrowsAfter(5) for i in range(5)])
```

Comment:
I think this would be clearer as `[ThrowsAfter(5)]*10`

As you have it written, it will throw in the second pass, not the first pass!

Reply:
Thanks for catching that!